player/player.go

```diff
@@ -0,0 +1,227 @@
+	SDHash      []byte          `json:"stream_hash"`
+	Blobs       []reflectedBlob `json:"blobs"`
+	Key         string          `json:"key"`
+	DecodedKey  []byte          `json:"-"`
+	Size        int             `json:"-"`
+	ContentType string          `json:"-"`
+}
+
+type reflectedBlob struct {
+	Hash          string `json:"blob_hash"`
+	Number        int    `json:"blob_num"`
+	IV            string `json:"iv"`
+	DecodedIV     []byte `json:"-"`
+	Length        int    `json:"length"`
+	encryptedBody []byte
+}
+
+// PlayURI downloads and streams LBRY video content located at uri delimited by rangeHeader
+// (use rangeHeader := request.Header.Get("Range"))
+func PlayURI(uri string, rangeHeader string, writer http.ResponseWriter) (err error) {
+	start, end := parseRange(rangeHeader)
+	stream := newReflectedStream(uri)
+	err = stream.resolve()
+	if err != nil {
+		return err
+	}
+	err = stream.fetchData()
+	if err != nil {
+		return err
+	}
+	blobStart, blobEnd := stream.prepareWriter(start, end, writer)
+	err = stream.streamBlobs(blobStart, blobEnd, writer)
+	return err
+}
+
+func parseRange(header string) (int64, int64) {
+	r := regexp.MustCompile(`bytes=(\d+)-(\d*)`)
+	m := r.FindStringSubmatch(header)
+	if len(m) == 0 {
+		return 0, 0
+	}
+	start, err1 := strconv.ParseInt(m[1], 10, 64)
+	end, err2 := strconv.ParseInt(m[2], 10, 64)
+	if err1 != nil || err2 != nil {
+		return 0, 0
+	}
+	if start < 0 {
+		start = 0
+	}
+	if end < 0 {
+		end = 0
+	}
+	if start > end {
+		start = 0
+		end = 0
+	}
+	return start, end
+}
+
+func newReflectedStream(uri string) *reflectedStream {
+	return &reflectedStream{URI: uri}
+}
+
+func (s *reflectedStream) URL() string {
+	return reflectorURL + string(s.SDHash)
+}
+
+// resolve gets SD hash from the daemon for a given URI to enable further retrieval of stream content from
+// reflector servers
+func (s *reflectedStream) resolve() error {
+	client := ljsonrpc.NewClient(config.Settings.GetString("Lbrynet"))
+	response, err := client.Resolve(s.URI)
+	if err != nil {
+		return err
+	}
+	source := (*response)[s.URI].Claim.Value.Stream.Source
+	s.SDHash = source.Source
+	s.ContentType = source.GetContentType()
+	monitor.Logger.WithFields(log.Fields{
+		"sd_hash":      fmt.Sprintf("%s", s.SDHash),
+		"uri":          s.URI,
+		"content_type": s.ContentType,
+	}).Info("resolved uri")
+	return nil
+}
+
+// fetchData gets stream data from the reflector and does some blob magic along the way
+func (s *reflectedStream) fetchData() error {
+	if s.SDHash == nil {
+		return errors.Err("No hash set, call `resolve` first")
+	}
+	monitor.Logger.WithFields(log.Fields{
+		"uri": s.URI,
+		"url": s.URL(),
+	}).Info("requesting stream data")
+	httpRequest, err := http.NewRequest("GET", s.URL(), new(bytes.Buffer))
+	httpClient := &http.Client{}
+	httpResponse, err := httpClient.Do(httpRequest)
+	if err != nil {
+		return err
+	}
+	defer httpResponse.Body.Close()
+	err = json.NewDecoder(httpResponse.Body).Decode(&s)
```

Comment:
I recommend using the types in lbry.go/stream as much as possible. we already have types defined for an SDBlob, and other useful parts.

Instead of doing a lot of the stuff in this function, you read httpResponse.Body into a byte slice, then do

```
sd := &stream.SDBlob{}
sd.FromBlob(data)
```

Now you have a proper sd blob. You can iterate through the BlobInfos and add those up to get the size. Or even better, you could add a function to SDBlob that takes a startBytes and endBytes and returns which blobs those bytes are in.